src/compiler.ts

```diff
@@ -3812,6 +3812,46 @@ export class Compiler extends DiagnosticEmitter {
       : expr;
   }
 
+  private convertFloat32ToBoolExpression(expr: ExpressionRef): ExpressionRef {
+    var module = this.module;
+    // 0 < abs(bitCast(x)) <= bitCast(Infinity) or
+    // (reinterpret<u32>(x) & 0x7FFFFFFF) - 1 <= 0x7F800000 - 1
+    //
+    // and finally:
+    // (reinterpret<u32>(x) << 1) - (1 << 1) <= ((0x7F800000 - 1) << 1)
+    expr = module.binary(BinaryOp.LeU32,
+      module.binary(BinaryOp.SubI32,
+        module.binary(BinaryOp.ShlI32,
+          module.unary(UnaryOp.ReinterpretF32, expr),
+          module.i32(1)
+        ),
+        module.i32(2) // 1 << 1
+      ),
+      module.i32(0xFEFFFFFE) // (0x7F800000 - 1) << 1
+    );
+    return expr;
+  }
+
+  private convertFloat64ToBoolExpression(expr: ExpressionRef): ExpressionRef {
+    var module = this.module;
+    // 0 < abs(bitCast(x)) <= bitCast(Infinity) or
+    // (reinterpret<u64>(x) & 0x7FFFFFFFFFFFFFFF) - 1 <= 0x7FF0000000000000 - 1
+    //
+    // and finally:
+    // (reinterpret<u64>(x) << 1) - (1 << 1) <= ((0x7FF0000000000000 - 1) << 1)
+    expr = module.binary(BinaryOp.LeU64,
+      module.binary(BinaryOp.SubI64,
+        module.binary(BinaryOp.ShlI64,
+          module.unary(UnaryOp.ReinterpretF64, expr),
+          module.i64(1)
+        ),
+        module.i64(2) // 1 << 1
+      ),
+      module.i64(0xFFFFFFFE, 0xFFDFFFFF) // (0x7FF0000000000000 - 1) << 1
+    );
+    return expr;
+  }
+
```

Comment:
Conceptionally the logic here is `makeIsTrueish` for the `f32` and `f64` cases. Might make sense to move it there...

Reply:
Good catch!